File: Tests/ConcatenatorTests.cs
using System;
using System.Text.Patterns;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace Tests {
	[TestClass]
	public class ConcatenatorTests {
		[TestMethod]
		public void Constructor() {
			Pattern _ = (Pattern)"Hello" & ' ' & "world";
		}

		[TestMethod]
		public void Consume() {
			Pattern HelloWorld = (Pattern)"Hello" & ' ' & "world";
			Pattern GoodbyeWorld = (Pattern)"Goodbye" & ' ' & "world";
			Result Result;

			Result = HelloWorld.Consume("Hello world");
			ResultAssert.Captures("Hello world", Result);

			Result = GoodbyeWorld.Consume("Goodbye world");
			ResultAssert.Captures("Goodbye world", Result);

			Result = HelloWorld.Consume("Hello everyone");
			ResultAssert.Fails(Result);

			Result = GoodbyeWorld.Consume("Hello world");
			ResultAssert.Fails(Result);
		}
	}
}

File: Tests/Patterns/ConcatenatorTests.fs
namespace Tests.Patterns

open System
open Stringier.Patterns
open Microsoft.VisualStudio.TestTools.UnitTesting

[<TestClass>]
type ConcatenatorTests() =

    [<TestMethod>]
    member _.``constructor`` () =
        let _pattern = "Hello" >> ' ' >> "world"
        ()

    [<TestMethod>]
    member _.``consume`` () =
        let helloWorld = "Hello" >> ' ' >> "world"
        let goodbyeWorld = "Goodbye" >> ' ' >> "world"

        let mutable result = Result()

        result <- helloWorld.Consume("Hello world")
        ResultAssert.Captures("Hello world", result)

        result <- goodbyeWorld.Consume("Goodbye world")
        ResultAssert.Captures("Goodbye world", result)

        result <- helloWorld.Consume("Hello everyone")
        ResultAssert.Fails(result)

        result <- goodbyeWorld.Consume("Hello world")
        ResultAssert.Fails(result)

    [<TestMethod>]
    member _.``forward composition still works`` () =
        let f x = x + 1
        let g x = x * 2
        let h = f >> g
        Assert.AreEqual(4, h 1)